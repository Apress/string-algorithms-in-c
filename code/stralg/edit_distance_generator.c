
#include "edit_distance_generator.h"
#include "cigar.h"

#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>


enum edit_op {
    RECURSE,
    INSERTION,
    DELETION,
    MATCH
};
struct deletion_info {
    char a;
};
struct match_info {
    char a;
};

struct edit_iter_frame {
    enum edit_op op;

    // The character we should delete or match
    uint8_t a;
    // Have we inserted or matched yet?
    bool at_beginning;
    
    // Fronts of buffers
    const uint8_t *pattern_front;
    uint8_t *string_front;
    char *cigar_front;
    
    // Number of edits left
    int max_dist;
    
    // The rest of the stack
    struct edit_iter_frame *next;
};


// I am heap allocating the links in the
// stack here. It might be faster to have an
// allocation pool.
static struct edit_iter_frame *
push_edit_iter_frame(
    enum edit_op op,
    bool at_beginning,
    const uint8_t *pattern_front,
    uint8_t *string_front,
    char *cigar_front,
    int max_dist,
    struct edit_iter_frame *next
) {
    struct edit_iter_frame *frame =
        malloc(sizeof(struct edit_iter_frame));
    frame->op = op;
    frame->at_beginning = at_beginning;
    frame->pattern_front = pattern_front;
    frame->string_front = string_front;
    frame->cigar_front = cigar_front;
    frame->max_dist = max_dist;
    frame->next = next;
    return frame;
}

void init_edit_iter(
    struct edit_iter *iter,
    const uint8_t *pattern,
    const char *alphabet,
    int max_edit_distance
) {
    uint32_t n = 2 * (uint32_t)strlen((char *)pattern);

    iter->pattern = pattern;
    iter->alphabet = alphabet;

    iter->string = malloc(n); iter->string[n - 1] = '\0';
    iter->edits = malloc(n);  iter->edits[n - 1] = '\0';
    iter->cigar = malloc(n);

    iter->frames = push_edit_iter_frame(
        RECURSE,
        true,
        iter->pattern,
        iter->string,
        iter->edits,
        max_edit_distance,
        0
    );
}




bool next_edit_pattern(
    struct edit_iter *iter,
    struct edit_pattern *result
) {
    assert(iter);
    assert(result);

    if (iter->frames == 0) return false;

    // pop top frame
    struct edit_iter_frame *frame = iter->frames;
    iter->frames = frame->next;

    const uint8_t *pattern = frame->pattern_front;
    uint8_t *buffer = frame->string_front;
    char *cigar = frame->cigar_front;

    if (*pattern == '\0') {
        // No more pattern to match ...
        // terminate the buffer report
        *buffer = '\0';
        *cigar = '\0';
        edits_to_cigar(iter->cigar, iter->edits);
        result->pattern = iter->string;
        result->cigar = iter->cigar;
        free(frame);
        return true;

    } else if (frame->max_dist == 0) {
        // We can't edit any more, so just move
        // pattern to the string and call back
        uint32_t rest = (uint32_t)strlen((char *)pattern);
        for (uint32_t i = 0; i < rest; ++i) {
              buffer[i] = pattern[i];
              cigar[i] = 'M';
        }
        buffer[rest] = cigar[rest] = '\0';
        edits_to_cigar(iter->cigar, iter->edits);
        result->pattern = iter->string;
        result->cigar = iter->cigar;
        free(frame);
        return true;
    }

    switch (frame->op) {
        case RECURSE:
            for (const char *a = iter->alphabet; *a; a++) {
                if (!frame->at_beginning) {
                    iter->frames = push_edit_iter_frame(
                        DELETION,
                        false,
                        frame->pattern_front,
                        frame->string_front,
                        frame->cigar_front,
                        frame->max_dist,
                        iter->frames
                    );
                    iter->frames->a = *a;
                }
                iter->frames = push_edit_iter_frame(
                    MATCH,
                    false,
                    frame->pattern_front,
                    frame->string_front,
                    frame->cigar_front,
                    frame->max_dist,
                    iter->frames
                );
                iter->frames->a = *a;
            }
            iter->frames = push_edit_iter_frame(
                INSERTION,
                false,
                frame->pattern_front,
                frame->string_front,
                frame->cigar_front,
                frame->max_dist,
                iter->frames
            );
            break;

        case INSERTION:
            *cigar = 'I';
            iter->frames = push_edit_iter_frame(
                RECURSE,
                false,
                frame->pattern_front + 1,
                frame->string_front,
                frame->cigar_front + 1,
                frame->max_dist - 1,
                iter->frames
            );
            break;

        case DELETION:
            if (frame->at_beginning) break;
            *buffer = frame->a;
            *cigar = 'D';
            iter->frames = push_edit_iter_frame(
                RECURSE,
                false,
                frame->pattern_front,
                frame->string_front + 1,
                frame->cigar_front + 1,
                frame->max_dist - 1,
                iter->frames
            );

            break;
        case MATCH:
            if (frame->a == *pattern) {
                *buffer = frame->a;
                *cigar = 'M';
                iter->frames = push_edit_iter_frame(
                    RECURSE,
                    false,
                    frame->pattern_front + 1,
                    frame->string_front + 1,
                    frame->cigar_front + 1,
                    frame->max_dist,
                    iter->frames
                );
            } else {
                *buffer = frame->a;
                *cigar = 'M';
                iter->frames = push_edit_iter_frame(
                    RECURSE,
                    false,
                    frame->pattern_front + 1,
                    frame->string_front + 1,
                    frame->cigar_front + 1,
                    frame->max_dist - 1,
                    iter->frames
                );
            }
            break;

        default:
            assert(false); // LCOV_EXCL_LINE
    }

    free(frame);
    return next_edit_pattern(iter, result); // recurse...
}

void dealloc_edit_iter(struct edit_iter *iter)
{
    struct edit_iter_frame *frame = iter->frames;
    while (frame) {
        struct edit_iter_frame *next = frame->next;
        free(frame);
        frame = next;
    }
    
    free(iter->string);
    free(iter->edits);
    free(iter->cigar);
}









void report(uint8_t *x, char *y) {
    // nop
}

void recursive_generator(
    const uint8_t *pattern_front,
    const uint8_t *alphabet,
    // To avoid initial deletions
    bool at_beginning,
    // Write the edited string here
    uint8_t *string_front,
    // Holds the beginning of full buffer
    // so we can report the string
    uint8_t *string,
    // We write the output cigar here
    char *cigar,
    // We build the edit string here
    char *edits_front,
    // and use the beggining of the edits buffer
    // when we report
    char *edits,
    int max_edit_distance)
{
    if (*pattern_front == '\0') {
        // No more pattern to match ...
        // Terminate the buffer and report
        *string_front = '\0';
        *edits_front = '\0';
        edits_to_cigar(cigar, edits);
        report(string, cigar);
        
    } else if (max_edit_distance == 0) {
        // We can't edit any more, so just move the
        // pattern to buffer and report
        uint32_t rest = (uint32_t)strlen((char *)pattern_front);
        for (uint32_t i = 0; i < rest; ++i) {
            string_front[i] = pattern_front[i];
            edits_front[i] = 'M';
        }
        string_front[rest] = cigar[rest] = '\0';
        edits_to_cigar(cigar, edits);
        report(string, cigar);
        
    } else {
        // RECURSION
        // Insertion
        *edits_front = 'I';
        recursive_generator(pattern_front + 1,
                            alphabet,
                            false,
                            string_front, string,
                            cigar,
                            edits_front + 1, edits,
                            max_edit_distance - 1);
        // Deletion
        if (!at_beginning) {
            for (const uint8_t *a = alphabet; *a; a++) {
                *string_front = *a;
                *edits_front = 'D';
                recursive_generator(pattern_front,
                                    alphabet,
                                    at_beginning,
                                    string_front + 1,
                                    string,
                                    cigar,
                                    edits_front + 1, edits,
                                    max_edit_distance - 1);
            }
        }
        // Match / substitution
        for (const uint8_t *a = alphabet; *a; a++) {
            if (*a == *pattern_front) {
                *string_front = *a;
                *edits_front = 'M';
                recursive_generator(pattern_front + 1,
                                    alphabet,
                                    false,
                                    string_front + 1,
                                    string,
                                    cigar,
                                    edits_front + 1, edits,
                                    max_edit_distance);
            } else {
                *string_front = *a;
                *edits_front = 'M';
                recursive_generator(pattern_front + 1,
                                    alphabet,
                                    false,
                                    string_front + 1,
                                    string,
                                    cigar,
                                    edits_front + 1, edits,
                                    max_edit_distance - 1);
            }
        }
    }
}

