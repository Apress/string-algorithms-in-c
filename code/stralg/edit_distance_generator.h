
#ifndef EDIT_DISTANCE_GENERATOR_H
#define EDIT_DISTANCE_GENERATOR_H

#include <stddef.h>
#include <stdbool.h>
#include <stdint.h>


/**
 Iterator for exploring an edit cloud.
 
 This iterator iterates through all strings in a given
 edit distance from a core. Consider the structure
 opaque; it is public so it can be stack allocated.
 */
struct edit_iter {
    const uint8_t *pattern;
    const char *alphabet;

    uint8_t *string;
    char *edits;
    char *cigar;

    struct edit_iter_frame *frames;
};
/**
 The result when the iterator gives a value.
 
 the pattern and cigar strings give you the string after
 editing from the core and the cigar that explains the
 edits. You must copy them if you need them; they will
 be updated next time you progress the iterator.
 */
struct edit_pattern {
    const uint8_t *pattern;
    const char *cigar;
};

/**
 Initialise an edit iterator.

 Do not modify the pattern between when  you
 create an iterator and you deallocate it again.
 This will break the iterator.
 
 @param iter the iterator
 @param core the core string to build edits around
 @param alphabet the alphabet used for inserting letters
 @param max_edit_distance it will explore strings up to
 this edit distance.
 */
void init_edit_iter(
    struct edit_iter *iter,
    const uint8_t *core,
    const char *alphabet,
    int max_edit_distance
);

/**
 Progress the iterator to the next edited string and provided the result.
 
 @param iter the iterator
 @param result the result structure that will be filled out
 with the data for the next string if this function returns true.
 @return true if there are more strings and false otherwise
 */
bool next_edit_pattern(
    struct edit_iter *iter,
    struct edit_pattern *result
);

/**
 Deallocate the resources held by an iterator.
 
 This function frees the resources stored in the
 iterator structor, not the struct itself. If it is
 heap allocated you are responsible for freeing
 the structure.
 
 @param iter the iterator to deallocate.
 */
void dealloc_edit_iter(
    struct edit_iter *iter
);

#endif

