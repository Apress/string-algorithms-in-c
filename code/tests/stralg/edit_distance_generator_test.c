
#include <edit_distance_generator.h>

#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdio.h>

int main(int argc, char * argv[])
{
    const uint8_t *pattern = (uint8_t *)"acca";
    const char *alphabet = "acgt";

    struct edit_iter iter;
    struct edit_pattern result;
    init_edit_iter(&iter, pattern, alphabet, 1);
    while (next_edit_pattern(&iter, &result)) {
        printf("%s %s\n",
        result.pattern, result.cigar);
    }
    dealloc_edit_iter(&iter);

    return EXIT_SUCCESS;
}
