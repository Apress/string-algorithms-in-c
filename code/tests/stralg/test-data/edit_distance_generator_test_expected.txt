cca 1I3M
tcca 4M
gcca 4M
ccca 4M
aca 1M1I2M
atca 4M
atcca 1M1D3M
agca 4M
agcca 1M1D3M
aca 2M1I1M
acta 4M
actca 2M1D2M
acga 4M
acgca 2M1D2M
acc 3M1I
acct 4M
accta 3M1D1M
accg 4M
accga 3M1D1M
accc 4M
accca 3M1D1M
acca 4M
accaa 3M1D1M
accca 2M1D2M
acaa 4M
acaca 2M1D2M
accca 1M1D3M
aaca 4M
aacca 1M1D3M
